lib/RecipeExecutor.jsm

```diff
@@ -0,0 +1,1075 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {tokenize} = ChromeUtils.import("resource://activity-stream/lib/Tokenize.jsm", {});
+
+/**
+ * RecipeExecutor is the core feature engineering pipeline for the in-browser
+ * personalization work. These pipelines are called "recipes". A recipe is an
+ * array of objects that define a "step" in the recipe. A step is simply an
+ * object with a field "function" that specifies what is being done in the step
+ * along with other fields that are semantically defined for that step.
+ *
+ * There are two types of recipes "builder" recipes and "combiner" recipes. Builder
+ * recipes mutate an object until it matches some set of critera. Combiner
+ * recipes take two objects, (a "left" and a "right"), and specify the steps
+ * to merge the right object into the left object.
+ *
+ * A short nonsense example recipe is:
+ * [ {"function": "get_url_domain", "path_length": 1, "field": "url", "dest": "url_domain"},
+ *   {"function": "nb_tag", "fields": ["title", "description"]},
+ *   {"function": "conditionally_nmf_tag", "fields": ["title", "description"]} ]
+ *
+ * Recipes are sandboxed by the fact that the step functions must be explicitly
+ * whitelisted. Functions whitelisted for builder recipes are specifed in the
+ * RecipeExecutor.ITEM_BUILDER_REGISTRY, while combiner functions are whitelisted
+ * in RecipeExecutor.ITEM_COMBINER_REGISTRY .
+ */
+this.RecipeExecutor = class RecipeExecutor {
+  constructor(nbTaggers, nmfTaggers) {
+    this.ITEM_BUILDER_REGISTRY = {
+      nb_tag: this.naiveBayesTag,
+      conditionally_nmf_tag: this.conditionallyNmfTag,
+      accept_item_by_field_value: this.acceptItemByFieldValue,
+      tokenize_url: this.tokenizeUrl,
+      get_url_domain: this.getUrlDomain,
+      tokenize_field: this.tokenizeField,
+      copy_value: this.copyValue,
+      keep_top_k: this.keepTopK,
+      scalar_multiply: this.scalarMultiply,
+      elementwise_multiply: this.elementwiseMultiply,
+      vector_multiply: this.vectorMultiply,
+      scalar_add: this.scalarAdd,
+      vector_add: this.vectorAdd,
+      make_boolean: this.makeBoolean,
+      whitelist_fields: this.whitelistFields,
+      filter_by_value: this.filterByValue,
+      l2_normalize: this.l2Normalize,
+      prob_normalize: this.probNormalize,
+      set_default: this.setDefault,
+      lookupValue: this.lookupValue,
+      copy_to_map: this.copyToMap,
+      scalar_multiply_tag: this.scalarMultiplyTag,
+      apply_softmax_tags: this.applySoftmaxTags
+    };
+    this.ITEM_COMBINER_REGISTRY = {
+      combiner_add: this.combinerAdd,
+      combiner_max: this.combinerMax,
+      combiner_collect_values: this.combinerCollectValues
+    };
+    this.nbTaggers = nbTaggers;
+    this.nmfTaggers = nmfTaggers;
+  }
+
+  /**
+   * Determines the type of a field. Valid types are:
+   *   string
+   *   number
+   *   array
+   *   map (strings to anything)
+   */
+  _typeOf(data) {
```

Comment:
How about other types? Such as function, boolean, regex, etc.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures

Reply:
We only really need to support the stuff that can come across JSON. I'll add null and boolean.

Reply:
Cool, let's comment the scope of this function then, for the readability.